File: 2021/deno/day-9.ts
const input = await Deno.readTextFile("./input/day-9.in");

// const input = await Deno.readTextFile("./input/example.in");

const rows = input.split("\n");

const height = rows.length;
const width = rows[0].length;

const grid = rows.map((row) => row.split("").map(Number));

const adj = calcAdj(width, height);

const origins = [];

for (let y = 0; y < height; y++) {
  for (let x = 0; x < width; x++) {
    const value = grid[y][x];

    const adjacent = adj[norm(x, y, width)].map((v) => {
      const { x, y } = invNorm(v, width);
      return grid[y][x];
    });

    const isLow = adjacent.every((v) => v > value);

    if (isLow) {
      origins.push({ x, y });
    }
  }
}

/**
 * Part One
 */
console.log(
  "Part One:",
  origins.map(({ x, y }) => grid[y][x] + 1).reduce((a, b) => a + b, 0)
);

/**
 * Part Two
 */

const flatGrid = flatten2dMatrix(grid);

const basins = [];

for (const { x, y } of origins) {
  const distances = bfs(norm(x, y, width), adj, flatGrid, width * height);
  const size = distances.filter((x) => x > 0).length + 1;

  basins.push(size);
}

basins.sort((a, b) => b - a);

console.log(
  "Part Two:",
  basins.slice(0, 3).reduce((a, b) => a * b, 1)
);

function norm(x: number, y: number, width: number) {
  return x + y * width;
}

function invNorm(normal: number, width: number) {
  return { x: normal % width, y: Math.floor(normal / width) };
}

function createMatrix<T>(size: number, init: () => T): T[] {
  return Array.from({ length: size }, init);
}

function flatten2dMatrix<T>(matrix: T[][]): T[] {
  const result: T[] = [];

  for (let y = 0; y < matrix.length; y++) {
    const row = matrix[y];
    result.push(...row);
  }

  return result;
}

function calcAdj(width: number, height: number) {
  const adj = createMatrix<number[]>(width * height, () => []);

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let index = norm(x, y, width);
      if (x > 0) {
        // down
        adj[index].push(norm(x - 1, y, width));
      }
      if (y + 1 < height) {
        // right
        adj[index].push(norm(x, y + 1, width));
      }
      if (x + 1 < width) {
        // left
        adj[index].push(norm(x + 1, y, width));
      }
      if (y > 0) {
        // up
        adj[index].push(norm(x, y - 1, width));
      }
    }
  }

  return adj;
}

export function bfs(
  start: number,
  adj: number[][],
  grid: number[],
  size: number
) {
  const q: number[] = [];
  const visited = createMatrix(size, () => false);
  const distance = createMatrix(size, () => 0);

  visited[start] = true;
  distance[start] = 0;
  q.push(start);

  while (true) {
    const current = q.shift();

    if (current == null) break;

    for (const vec of adj[current]) {
      if (visited[vec]) continue;

      visited[vec] = true;

      if (grid[vec] === 9) continue;

      distance[vec] = distance[current] + 1;

      q.push(vec);
    }
  }

  return distance;
}

File: 2021/swift/day-9.swift
import Foundation

let filename = "../input/day-9.in"

func coordToIndex(_ coords: (x: Int, y: Int), _ width: Int) -> Int {
    coords.y * width + coords.x
}

func indexToCoord(_ index: Int, _ width: Int) -> (x: Int, y: Int) {
    (index % width, index / width)
}

// create adjacent matrix
func calcAdj<T>(_ graph: [T], _ width: Int) -> [[Int]] {
    var adjacent: [[Int]] = []

    let height = graph.count / width

    for (index, _) in graph.enumerated() {
        var localAdj: [Int] = []
        let (x, y) = indexToCoord(index, width)

        let up = (x, y - 1)
        let down = (x, y + 1)
        let left = (x - 1, y)
        let right = (x + 1, y)

        if up.1 >= 0 {
            localAdj.append(coordToIndex(up, width))
        }

        if down.1 < height {
            localAdj.append(coordToIndex(down, width))
        }

        if left.0 >= 0 {
            localAdj.append(coordToIndex(left, width))
        }

        if right.0 < width {
            localAdj.append(coordToIndex(right, width))
        }

        adjacent.append(localAdj)
    }

    return adjacent
}

struct Queue<Element> {
    var items: [Element] = []
    mutating func push(_ item: Element) {
        items.append(item)
    }

    mutating func pop() -> Element {
        return items.removeLast()
    }

    mutating func front() -> Element {
        return items.remove(at: 0)
    }

    func empty() -> Bool {
        return items.count == 0
    }
}

// perform actual BFS
func bfs(_ root: Int, _ graph: [Int], _ adj: [[Int]]) -> [Int] {
    var queue = Queue<Int>()
    var visited = graph.map { _ in
        false
    }

    var members: [Int] = []

    visited[root] = true
    members.append(root)

    queue.push(root)

    while !queue.empty() {
        let next = queue.front()

        for node in adj[next] {
            if visited[node] { continue }

            if graph[node] == 9 {
                continue
            }

            visited[node] = true

            members.append(node)

            queue.push(node)
        }
    }

    return members
}

// turn a 2d graph into 1d
func flatGraph<T>(_ graph: [[T]]) -> [T] {
    var flat: [T] = []

    for row in graph {
        flat.append(contentsOf: row)
    }

    return flat
}

func prettyPrint<T>(_ graph: [T], _ width: Int, format: (T) -> String) {
    for lead in 0 ..< graph.count / width {
        let from = lead * width
        let to = from + width - 1
        let row = graph[from ... to].map { format($0) }
        print(row.joined(separator: " "))
    }
}

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let grid = contents.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: "\n").map {
            row in
            row.split(separator: "").map { Int($0)! }
        }

        let width = grid[0].count

        let graph = flatGraph(grid)

        let adj = calcAdj(graph, width)

        let partOne = graph.enumerated().reduce(0) {
            acc, curr in
            let (index, node) = curr

            return acc + (adj[index].allSatisfy { graph[$0] > node } ? 1 + node : 0)
        }

        print("Part one:", partOne)

        let basins: [Int] = graph.enumerated().filter {
            curr in
            let (index, node) = curr

            return adj[index].allSatisfy { graph[$0] > node }
        }.map { $0.0 }

        var groups = basins.reduce([[Int]]()) {
            acc, basin in
            var next = acc
            next.append(bfs(basin, graph, adj))
            return next
        }

        groups.sort(by: { $0.count > $1.count })

        let partTwo = groups[0 ... 2].map { $0.count }.reduce(1, *)

        print("Part two:", partTwo)

    } catch { print(error) }
}

main()
